### PR TITLE
Discriminative margin based clustering loss fix

### PR DESCRIPTION
**Pull request by dBeker**
Current discriminative margin based clustering loss contains several implementation issues:
- Errors at documentation
- Errors at the usage of embeddings and wrong binding of embedding dimensions into instance numbers. 

This PR aims to fix those issues.

### CODE REVIEW

File: q! [hunk 1]
```diff
@@ -0,0 +1,31 @@
+* [32mdiscriminative_margin_based_clustering_loss_fix[m
```
Comment (tkerola) [inline q!]: This file should be deleted.

File: chainer/functions/loss/discriminative_loss.py [hunk 2]
```diff
@@ -12,12 +15,12 @@ class DiscriminativeMarginBasedClusteringLoss(object):
 
     This is the implementation of the following paper:
     https://arxiv.org/abs/1708.02551
-    This method is a semi-supervised solution to instance segmentation.
     It calculates pixel embeddings, and calculates three different terms
     based on those embeddings and applies them as loss.
-    The main idea is that the pixel embeddings
-    for same instances have to be closer to each other (pull force),
-    for different instances, they have to be further away (push force).
+    This loss is penalizes the pixel embeddings according to following items:
```
Comment (tkerola) [inline chainer/functions/loss/discriminative_loss.py]: Should be: `This loss penalizes`

File: chainer/functions/loss/discriminative_loss.py [hunk 3]
```diff
@@ -77,65 +76,85 @@ def __call__(self, embeddings, labels):
...
+        # Find active labels per batch item
+        active_id_count = xp.zeros((shape[0], ))
+        active_idxs = []
+        for b_idx in range(shape[0]):
+            active_id = xp.unique(labels[b_idx])
+            active_id = active_id[active_id >= 0]
+            active_idxs.append(active_id)
+            active_id_count[b_idx] = len(active_id)
+
+        # Calculate mean embeddings ans variance loss
```
Comment (tkerola) [inline chainer/functions/loss/discriminative_loss.py]: `ans` -> `and`

File: chainer/functions/loss/discriminative_loss.py [hunk 4]
```diff
@@ -77,65 +76,85 @@ def __call__(self, embeddings, labels):
...
-            # Old numpy does not have numpy.stack.
-            ms = xp.concatenate([xp.expand_dims(x, 0) for x in ms], 0)
-        mns = c_sum(emb * ms, axis=(3, 4))
-        mns = mns / xp.maximum(xp.sum(ms, (2, 3, 4))[:, :, None], 1)
-        mns_exp = mns[:, :, :, None, None]
+        unique_label_idx = xp.unique(labels)
+        # Remove background label
+        unique_label_idx = unique_label_idx[unique_label_idx >= 0]
+
+        var_loss = xp.zeros((shape[0],))
```
Comment (tkerola) [inline chainer/functions/loss/discriminative_loss.py]: How about improving readability by
```
b = embeddings.shape[0]
```
and then using `b` throughout the code to denote the batch size?
```
xp.zeros(b)
```

Comment (tkerola) [reply chainer/functions/loss/discriminative_loss.py]: I think `var_loss` should have the same dtype as `embeddings`.

File: chainer/functions/loss/discriminative_loss.py [hunk 5]
```diff
@@ -77,65 +76,85 @@ def __call__(self, embeddings, labels):
...
+            local_var_loss = c_sum(local_var_loss, (1, 2))
+            dividend = xp.maximum(active_id_count, 1) * number_of_pixels[:, 0]
+            var_loss += local_var_loss / dividend
+
+        var_loss = average(var_loss)
+
+        # Calculate mean distance loss
+        means = stack(means, 1)
+
+        dist_loss = 0.0
```
Comment (tkerola) [inline chainer/functions/loss/discriminative_loss.py]: The dtype should be the same as `embeddings`.

**Comment by dBeker** [comment]
@tkerola Thank you. I fixed the issues that you mentioned.

File: chainer/functions/loss/discriminative_loss.py [hunk 6]
```diff
@@ -35,17 +38,12 @@ class DiscriminativeMarginBasedClusteringLoss(object):
     """
 
     def __init__(self, delta_v=0.5, delta_d=1.5,
-                 max_embedding_dim=10, norm=1,
-                 alpha=1.0, beta=1.0, gamma=0.001):
+                 norm=1, alpha=1.0, beta=1.0, gamma=0.001):
```
Comment (niboshi) [inline chainer/functions/loss/discriminative_loss.py]: Currently all the arguments can be specified as positional. If `max_embedding_dim` is to be removed, all the succeeding arguments must be keyword-only. Otherwise the user codes can be *silently* broken.

You can use `chainer.utils.argument.parse_kwargs` to make them keyword-only
(see `chainer.training.extensions.snapshot()` in `training/extensions/_snapshot.py` as an exapmle).

In that case, please update the docstring to show the "logical" signature.
`chainer.functions.convolution_2d` might be a good example to do that (see the first line of the docstring).

File: chainer/functions/loss/discriminative_loss.py [hunk 7]
```diff
@@ -1,5 +1,8 @@
 from chainer import backend
 from chainer.functions.activation.relu import relu
+from chainer.functions.math.average import average
+
```
Comment (niboshi) [inline chainer/functions/loss/discriminative_loss.py]: Please remove the blank line.

File: chainer/functions/loss/discriminative_loss.py (hunk 6 above)
Comment (niboshi) [reply chainer/functions/loss/discriminative_loss.py]: There's yet another complexity: deprecation period.

We have a policy to place a deprecation period of at least one major release.
During this period, existing user code with `max_embedding_dim`, or those giving the succeeding arguments as positional, should work just as before (with `DeprecationWarning`).

It can be a bit too tricky to support both the existing and new-style interface at the same time. I think I'd like to do that in a separate PR. In this PR, could you simply leave the argument `max_embedding_dim` and warn if the argument is explicitly given (either as positional or keyword)?

I mean, let the signature as below (note that the default value is `None`):

```py
def __init__(self, delta_v=0.5, delta_d=1.5,
             max_embedding_dim=None, norm=1,
             alpha=1.0, beta=1.0, gamma=0.001):
```

And warn if `max_embedding_dim` is given:

```py
if max_embedding_dim is not None:
    warnings.warn(
        'blah',
        DeprecationWarning)
```

We should also warn if the remaining arguments (`norm`, etc.) are given as positional. I'd like to do that in a separate PR.

File: tests/chainer_tests/functions_tests/loss_tests/test_discriminative_margin_based_clustering_loss.py [hunk 8]
```diff
@@ -17,48 +17,50 @@
     'beta': [1],
     'gamma': [0.001],
     'norm': [1],
-    'result_l_dist': [6.0776708],
```
Comment (tkerola) [inline tests/chainer_tests/functions_tests/loss_tests/test_discriminative_margin_based_clustering_loss.py]: How about keeping the original test as well? We need to make sure that this loss is backwards-compatible with the previous version.

Comment (dBeker) [reply tests/chainer_tests/functions_tests/loss_tests/test_discriminative_margin_based_clustering_loss.py]: Ah, as we change the calculation method, the values also have to be adjusted for tests not to fail.

File: chainer/functions/loss/discriminative_loss.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (dBeker) [reply chainer/functions/loss/discriminative_loss.py]: Done

File: chainer/functions/loss/discriminative_loss.py (hunk 6 above)
Comment (dBeker) [reply chainer/functions/loss/discriminative_loss.py]: I have added `**kwargs` option as you suggested. Would it be ok as-is or shall I also add deprecation warning?

Comment (niboshi) [reply chainer/functions/loss/discriminative_loss.py]: Thank you for the fix.
It looks the current version misses this point:

>If `max_embedding_dim` is to be removed, all the succeeding arguments must be keyword-only. Otherwise the user codes can be *silently* broken.

For example, if a user has written a code like this:
```
DiscriminativeMarginBasedClusteringLoss(0.5, 1.5, 10, 2)
```

He meant `delta_v=0.5, delta_d=1.5, max_embedding_dim=10, norm=1, alpha=1.0`.
After this fix, this would become `delta_v=0.5, delta_d=1.5, norm=10, alpha=2`, all silently.

We have to keep the behavior of the code like above intact, and at the same time prepare for making all the arguments (`norm`, `alpha`, ...) keyword-only.

**Comment by kmaehashi** [comment]
As Chainer v7 went to the maintenance phase, we decided to close this PR as newly introducing deprecations cannot be accepted. Thank you again for your contribution!
